chef/cookbooks/cpe_activedirectory/resources/cpe_activedirectory.rb

```diff
@@ -8,70 +8,213 @@
+    end
+
+    node.default['cpe_profiles']["#{prefix}.active_directory"] = ad_profile
+
+    unless node.ad_reachable?(node['cpe_activedirectory']['bind_ldap_check_hostname'])
+      log 'cpe_activedirectory cannot communicate to domain - profile will fail to install' do
+        level :warn
+      end
+    end
+  end
+
+  def configure
+    # Force this into *resource* runtime, not compile time - otherwise this
+    # fails after immediate binds due to already_set being nil
+    # https://github.com/facebook/chef-utils/blob/master/Compile-Time-Run-Time.md
+    ruby_block 'Force code into resource runtime' do
+      block do
+        bind_options = node['cpe_activedirectory']['bind_options']
+
+        flags_to_run = []
+
+        node['cpe_activedirectory']['options'].each do |root, settings|
+          next unless [root, 'all'].include?(new_resource.scope)
+
+          settings.reject { |_k, v| v.nil? }.each do |key, value|
+            already_set = already_set?(node.active_directory_state, root, key, value)
+            if already_set.nil?
+              log "cpe_activedirectory option #{key} is not supported" do
+                level :warn
+              end
+              next
+            elsif !already_set
+              flags_to_run << "-#{key} #{shell_format(value)}"
+            end
+          end
+        end
+        flags_to_run.each do |flag|
+          cmd = "/usr/sbin/dsconfigad #{flag}"
+          if node['cpe_activedirectory']['what_if_execution']
+            log "Would have run '#{cmd}'" do
+              level :info
+            end
+          else
+            execute "Setting Active Directory value #{flag}" do
+              command cmd
+              only_if { node.ad_bound?(bind_options['HostName']) }
+            end
+          end
+        end
       end
     end
   end
 
-  flags_to_run.each do |flag|
-    cmd = "/usr/sbin/dsconfigad #{flag}"
-    if node['cpe_activedirectory']['what_if_execution']
-      log "Would have run '#{cmd}'" do
-        level :info
+  def shell_format(value)
+    if value.is_a?(Array)
+      "'#{value.join(',')}'"
+    elsif [true, false].include?(value)
+      value ? 'enable' : 'disable'
+    else
+      value.to_s
+    end
+  end
+
+  def unbind
+    unless node.ad_reachable?(node['cpe_activedirectory']['bind_ldap_check_hostname'])
+      log 'cpe_activedirectory cannot communicate to domain - will not attempt to force unbind' do
+        level :warn
+        return
       end
+    end
+
+    if node['cpe_activedirectory']['unbind_method'] == 'profile_resource'
+      unbind_profile
     else
-      cmd = Mixlib::ShellOut.new(cmd)
-      cmd.run_command
-      cmd.error!
+      unbind_execute
+    end
+  end
+
+  def unbind_execute
+    username_to_check = node['cpe_activedirectory']['bind_options']['UserName']
+    unless username_to_check.nil?
+      execute 'Force unbinding to domain' do
+        command '/usr/sbin/dsconfigad -force -remove -u cpe -p fakepass'
```

Comment:
User "cpe", password "fakepass" - is this filler, or is this meant to be hardcoded? If we need a username and password to unbind via dsconfigad, it should be exposed as an attribute as with binding.

Reply:
When you pass -force with unbind the credentials don't have to be real. It's undocumented except for an article by Rich Trouton.

Reply:
I love macOS too!